ops/relation.py

```diff
@@ -0,0 +1,548 @@
+    a relation with the provider charm has *not* been successful.
+    """
+    def __init__(self, handle, data=None):
+        super().__init__(handle)
+        self.data = data
+
+    def snapshot(self):
+        return {"data": self.data}
+
+    def restore(self, snapshot):
+        self.data = snapshot["data"]
+
+
+class ProviderUnready(EventBase):
+    """Event triggered when a provider is not ready
+
+    If a provider charm is not ready to service requests, when a
+    consumer charm for a new relation with it, or is already related
+    to it, then a :class:`ProviderUnready` event is raised. This
+    presumes that the provider charm as set its `ready` status to
+    `False`.
+
+    The :class:`ProviderUnready` event is raised regardless of wether
+    the provider charm is compatible or not. Compatibility checks are
+    done only if the provider charm is ready to service requests.
+    """
+    pass
+
+
+class ProviderBroken(EventBase):
+    """Event raised when provider consumer relation is disolved
+
+    If the realtion between a provider and consumer charm is removed,
+    then a :class:`ProviderBroken` relation is raised.
+    """
+    pass
+
+
+class ConsumerEvents(CharmEvents):
+    """Descriptor for consumer charm events
+    """
+    available = EventSource(ProviderAvailable)
+    invalid = EventSource(ProviderInvalid)
+    unready = EventSource(ProviderUnready)
+    broken = EventSource(ProviderBroken)
+
+
+class Consumer(Object):
+    """Manages relations with a resource provider
+
+    The :class:`Consumer` object manages relations with resource
+    provider charms, by checking compatibility between consumer
+    requirements and provider type and version specification. Any
+    charm that uses resources provided by other charms may manage its
+    relation with the providers by instantiating on :class:`Consumer`
+    object for each such relation. A :class:`Consumer` object may be
+    instantiated in the `__init__` method of the consumer charm as
+    follows
+
+    Example::
+
+        self.provider_name = Consumer(self, resource_name, consumes)
+
+    In managing the relation between provider and consumer, the
+    :class:`Consumer` object may raise any of the following events,
+    that a consumer charm can choose to respond to
+
+    - :class:`ProviderAvailable`
+    - :class:`ProviderInvalid`
+    - :class:`ProviderUnready`
+    - :class:`ProviderBroken`
+
+    Args:
+        charm: :class:`ops.charm.CharmBase` derived object that is
+            instantiating the :class:`Consumer` object. This is almost
+            always `self`.
+        name: string name of relation (as defined in `metadata.yaml`) that
+            consumer charms will use to interact with provider charms.
+        consumes: a dictionary containing acceptable provider
+            specifications. The dictonary must contain at least one
+            key `consumes`. The value of this key is itself a
```

Comment:
why is 'consumes' the only required sub field?
Why wouldn't this be:
`consumes={'mysql': '>5.0.2', 'mariadb': '<=6.1.0'}`

It seems odd to pass a dict that contains a dict with only 1 key which is the name of the parameter that you passed to this object.

Reply:
agreed

Reply:
The reason the top level data structure for "provides" and "consumes" was chosen to be a mapping (hence with only one key at present) is to support extensibility, both within the operator framework and also by derived objects. So for instance the current minimal implementation is only handing constraint checking on "service version" and "service type". This may be extended in arbitrary ways in the future as we discover more use cases for constraint checking on relations. In its most sophisticated incarnation such constraint checking may evolve into a general [Constraint Satisfaction Problem](https://en.wikipedia.org/wiki/Constraint_satisfaction_problem). 

I am open to suggestions on alternative names for the keys. For example will changing keys names to some thing like "service_specification" or some such that make it feel more natural, with the understanding that future revisions and derived implementations, may have additional constraints, other than "service_specification" . Alternatively would it help if I documented this rationale for the choice of a mapping as the top level data structure ?

Reply:
I have clarified this further in docstring of Provider and Consumer class, explaining that top level dictionary keys represent a conjunction of constraints.

Reply:
Nested dictionaries have been removed as discussed in review meeting dated 26 April 2021. Thank you for feedback.